internal/utils/tfsdk.go

```diff
@@ -0,0 +1,38 @@
+package utils
+
+import (
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+// SliceToListType converts a tfsdk naive []T1 into an types.List of []T2.
+// This handles both structs and simple types to attr.Values.
+func SliceToListType[T1 any, T2 any](ctx context.Context, value []T1, elemType attr.Type, path path.Path, diags diag.Diagnostics, iteratee func(item T1) T2) types.List {
```

Comment:
What is your hot take on helpers like this? While its not much of a problem in this data_source, I've had a ton of issues with naive slices as compared to types.List members. In the more complicated nested structs, these let you just inline a lot of boilerplate.

Reply:
👍 I think there's likely a lot of room for more helpers like this in this codebase.